File: Problem-1.java
// Time Complexity : O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach
// Perform a level-order traversal (BFS) of the binary tree using a queue.
// At each level, find the maximum value among all nodes in that level.
// Add the maximum value of each level to the result list.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        if (root == null){
            return new ArrayList<>();
        }
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        List<Integer> result = new ArrayList<>();

        while (!q.isEmpty()){
            int size = q.size();
            int max_value = Integer.MIN_VALUE;  
            for(int i =0; i <size; i++){
                TreeNode curr = q.poll();
                if (curr.val > max_value){
                    max_value =  curr.val;
                }
                if (curr.left != null){
                    q.offer(curr.left);
                }
                if (curr.right != null){
                    q.offer(curr.right);
                }
            }
            result.add(max_value);
        }
        return result;
        
    }
}

File: Problem-1.py
# Time Complexity : O(n)
# Space Complexity :O(n)
# Did this code successfully run on Leetcode :yes
# Any problem you faced while coding this :no


# Your code here along with comments explaining your approach
# Perform a level-order traversal (BFS) of the binary tree using a queue.
# At each level, find the maximum value among all nodes in that level.
# Add the maximum value of each level to the result list.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        from collections import deque
        if not root:
            return []
        queue = deque()
        output = []
        queue.append(root) #[1]
        while queue:
            size = len(queue)
            max_value = float("-inf")
            for i in range(size):
                curr = queue.popleft() #1
                if curr.val > max_value:
                    max_value = curr.val
                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
            output.append(max_value)
        return output
                

        